protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java

```diff
@@ -172,100 +172,136 @@ public void packetReceived(Session session, Packet packet) {
             } else if (packet instanceof ServerboundPingRequestPacket pingRequestPacket) {
                 session.send(new ClientboundPongResponsePacket(pingRequestPacket.getPingTime()));
             }
-        } else if (protocol.getState() == ProtocolState.GAME) {
+        } else if (protocol.getInboundState() == ProtocolState.GAME) {
             if (packet instanceof ServerboundKeepAlivePacket keepAlivePacket) {
-                if (keepAlivePacket.getPingId() == this.lastPingId) {
-                    long time = System.currentTimeMillis() - this.lastPingTime;
-                    session.setFlag(MinecraftConstants.PING_KEY, time);
+                KeepAliveState currentKeepAliveState = this.keepAliveState;
```

Comment:
What specifically is this being cached to a variable for? The field value seems to be mutated when updating the keep alive, but replaced on login and config start/end.

If we are trying to prevent races with the keep alive thread, have we considered making `KeepAliveState` immutable?

Also, could `keepAlivePending` be replaced with the field being null when not pending, and have the field be volatile?

Reply:
This is coded closer to mojangs impl.
The pending field and other fields are part of the commonpacketlistener. And each protocol state keeps track of separate keepalives.

Reply:
Changing it now would make diffing it with vanilla harder. Also, if I restructure it, there's the possibility something doesn't properly match up with vanilla.